### ISSUE TITLE
HttpClientFactory: resolving a typed client is significantly slower that resolving a named client.

### ISSUE THREAD

**Issue by damianh** [aspnet/Extensions#2060]
# Describe the bug

Resolving a TypedClient appears to have significant performance overhead in contrast to using a named client. The issue originally turned up on [an issue in ProxyKit](https://github.com/damianh/ProxyKit/issues/124). Discussion on [twitter](https://twitter.com/randompunter/status/1149604794618638336).

Code https://github.com/damianh/httpclientfactory-typed-client-di-perf/blob/3bbb62205b2fb9fa1862aad9b1a827d1224fd1b1/TypeClientBenchmark/Program.cs#L8-L28

``` ini

// * Summary *

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Core i7-3930K CPU 3.20GHz (Ivy Bridge), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.0.100-preview5-011568
  [Host]     : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT


|         Method |         Mean |        Error |      StdDev |
|--------------- |-------------:|-------------:|------------:|
| GetTypedClient | 247,299.6 ns | 5,268.002 ns | 5,855.37 ns |
| GetNamedClient |     341.4 ns |     7.065 ns |    20.72 ns |

// * Hints *
Outliers
  Program.GetTypedClient: Default -> 2 outliers were removed (277.79 us, 285.99 us)
  Program.GetNamedClient: Default -> 1 outlier  was  removed (402.63 ns)

// * Legends *
  Mean   : Arithmetic mean of all measurements
  Error  : Half of 99.9% confidence interval
  StdDev : Standard deviation of all measurements
  1 ns   : 1 Nanosecond (0.000000001 sec)

// ***** BenchmarkRunner: End *****
// ** Remained 0 benchmark(s) to run **
Run time: 00:01:36 (96.3 sec), executed benchmarks: 2

Global total time: 00:01:42 (102.64 sec), executed benchmarks: 2
```
### To Reproduce

Full repro here: https://github.com/damianh/httpclientfactory-typed-client-di-perf

### Expected behavior

Resolving typed clients should not have so much over head.

**Comment by rynowak** [comment]
> Resolving typed clients should not have so much over head.

Agreed, that's a shocking difference in performance.

**Comment by rynowak** [comment]
This is fixed for 3.0.0-preview8 we'll follow up on patching 2.2.X

**Comment by damianh** [comment]
@rynowak Was 2.2.x patched? Given that 2.1.x is LTS, I believe that also needs to be patched.
